File: cses/04_graph_algorithms/1193_labyrinth/solution.py
import sys
from collections import deque


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    m = int(data[1])
    grid = data[2:2 + n]

    # Bản đồ phẳng để truy cập O(1): flat[r*m + c] là ký tự của ô (r, c).
    flat = b"".join(grid)

    # Xác định chỉ số phẳng của ô xuất phát 'A' và ô đích 'B'.
    start = -1
    end = -1
    for r in range(n):
        row = grid[r]
        a = row.find(b"A")
        if a != -1:
            start = r * m + a
        b = row.find(b"B")
        if b != -1:
            end = r * m + b

    WALL = 35  # ord('#')

    # visited vừa đánh dấu đã thăm vừa lưu hướng để truy vết:
    # 0 = chưa thăm, 1=U, 2=D, 3=L, 4=R, 5 = ô xuất phát A.
    visited = bytearray(n * m)
    visited[start] = 5

    dq = deque()
    dq.append(start)
    append = dq.append
    popleft = dq.popleft

    # BFS từ 'A': mọi cạnh trọng số 1 nên đường đi tìm được là ngắn nhất theo số bước.
    while dq:
        cur = popleft()
        if cur == end:
            break
        r = cur // m
        c = cur - r * m

        # Xét 4 ô kề cạnh; ô nào trong lưới, không phải tường và chưa thăm thì thăm.
        if c > 0:
            nb = cur - 1
            if not visited[nb] and flat[nb] != WALL:
                visited[nb] = 3  # L
                append(nb)
        if c + 1 < m:
            nb = cur + 1
            if not visited[nb] and flat[nb] != WALL:
                visited[nb] = 4  # R
                append(nb)
        if r > 0:
            nb = cur - m
            if not visited[nb] and flat[nb] != WALL:
                visited[nb] = 1  # U
                append(nb)
        if r + 1 < n:
            nb = cur + m
            if not visited[nb] and flat[nb] != WALL:
                visited[nb] = 2  # D
                append(nb)

    # Nếu 'B' chưa được thăm thì không tồn tại đường đi.
    if not visited[end]:
        sys.stdout.buffer.write(b"NO\n")
        return

    letters = {1: b"U", 2: b"D", 3: b"L", 4: b"R"}
    # Offset để lùi về ô liền trước tùy theo hướng đã đi.
    back = {1: m, 2: -m, 3: 1, 4: -1}

    # Truy vết ngược từ 'B' về 'A' rồi đảo lại để có thứ tự từ 'A' tới 'B'.
    path = []
    cur = end
    while cur != start:
        code = visited[cur]
        path.append(letters[code])
        cur += back[code]
    path.reverse()

    sys.stdout.buffer.write(b"YES\n%d\n%s\n" % (len(path), b"".join(path)))


if __name__ == "__main__":
    main()

File: cses/04_graph_algorithms/1193_labyrinth/solution.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    // Đọc lưới và xác định chỉ số phẳng của ô xuất phát 'A' và ô đích 'B'.
    std::vector<std::string> grid(n);
    int start = -1;
    int target = -1;
    for (int row = 0; row < n; ++row) {
        std::cin >> grid[row];
        for (int column = 0; column < m; ++column) {
            if (grid[row][column] == 'A') {
                start = row * m + column;
            } else if (grid[row][column] == 'B') {
                target = row * m + column;
            }
        }
    }

    // previous_move[cell]: hướng đã dùng để đi vào ô đó, đồng thời đánh dấu đã thăm
    // (0 = chưa thăm, 'S' = ô xuất phát).
    std::vector<char> previous_move(n * m, 0);
    previous_move[start] = 'S';
    std::queue<int> queue;
    queue.push(start);

    constexpr int DR[4] = {-1, 1, 0, 0};
    constexpr int DC[4] = {0, 0, -1, 1};
    constexpr char MOVE[4] = {'U', 'D', 'L', 'R'};

    // BFS từ 'A': mọi cạnh trọng số 1 nên đường đi tìm được là ngắn nhất theo số bước.
    while (!queue.empty() && previous_move[target] == 0) {
        const int cell = queue.front();
        queue.pop();
        const int row = cell / m;
        const int column = cell % m;

        // Xét 4 ô kề cạnh; ô nào trong lưới, không phải tường và chưa thăm thì thăm.
        for (int direction = 0; direction < 4; ++direction) {
            const int nr = row + DR[direction];
            const int nc = column + DC[direction];
            if (nr < 0 || nr >= n || nc < 0 || nc >= m) {
                continue;
            }

            const int neighbor = nr * m + nc;
            if (grid[nr][nc] != '#' && previous_move[neighbor] == 0) {
                previous_move[neighbor] = MOVE[direction];
                queue.push(neighbor);
            }
        }
    }

    // Nếu 'B' chưa được thăm thì không tồn tại đường đi.
    if (previous_move[target] == 0) {
        std::cout << "NO\n";
        return 0;
    }

    // Truy vết ngược từ 'B' về 'A': đọc hướng đã lưu và lùi về ô liền trước.
    std::string path;
    for (int cell = target; cell != start;) {
        const char move = previous_move[cell];
        path.push_back(move);
        if (move == 'U') {
            cell += m;
        } else if (move == 'D') {
            cell -= m;
        } else if (move == 'L') {
            ++cell;
        } else {
            --cell;
        }
    }
    // Đảo ngược để có thứ tự đúng từ 'A' tới 'B'.
    std::reverse(path.begin(), path.end());

    std::cout << "YES\n" << path.size() << '\n' << path << '\n';
    return 0;
}
